### PR TITLE
PERF: Speed up secure media and ACL sync rake tasks

### PR DESCRIPTION
**Pull request by martin-brennan**
Incorporates learnings from /t/64227:

* Changes the code to set access control posts in the rake
  task to be an efficient UPDATE SQL query.
  The original version was timing out with 312017 post uploads,
  the new query took ~3s to run.
* Changes the code to mark uploads as secure/not secure in
  the rake task to be an efficient UPDATE SQL query rather than
  using UploadSecurity. This took a very long time previously,
  and now takes only a few seconds.
* Spread out ACL syncing for uploads into jobs with batches of
  100 uploads at a time, so they can be parallelized instead
  of having to wait ~1.25 seconds for each ACL to be changed
  in S3 serially.

One issue that still remains is post rebaking. Doing this serially
is painfully slow. We have a way to do this in sidekiq via PeriodicalUpdates
but this is limited by max_old_rebakes_per_15_minutes. It would
be better to fan this rebaking out into jobs like we did for the
ACL sync, but that should be done in another PR.

### CODE REVIEW

File: app/jobs/regular/sync_acls_for_uploads.rb [hunk 1]
```diff
@@ -18,7 +18,18 @@ def execute(args)
         Rails.logger.warn("Syncing ACL for upload ids: #{args[:upload_ids].join(", ")}")
         Upload.includes(:optimized_images).where(id: args[:upload_ids]).find_in_batches do |uploads|
           uploads.each do |upload|
-            Discourse.store.update_upload_ACL(upload, optimized_images_preloaded: true)
+            begin
+              Discourse.store.update_upload_ACL(upload, optimized_images_preloaded: true)
```
Comment (nbianca) [inline app/jobs/regular/sync_acls_for_uploads.rb:22]: Do you know if S3 api allows for bulk update of ACLs?

File: lib/tasks/uploads.rake [hunk 2]
```diff
@@ -589,73 +578,77 @@ task "uploads:secure_upload_analyse_and_update" => :environment do
...
-def adjust_acls(uploads_to_adjust_acl_for)
-  total_count = uploads_to_adjust_acl_for.respond_to?(:length) ? uploads_to_adjust_acl_for.length : uploads_to_adjust_acl_for.count
-  puts "", "Updating ACL for #{total_count} uploads.", ""
-  i = 0
-  uploads_to_adjust_acl_for.each do |upload|
-    RakeHelpers.print_status_with_label("Updating ACL for upload.......", i, total_count)
-    Discourse.store.update_upload_ACL(upload)
-    i += 1
+def adjust_acls(upload_ids_to_adjust_acl_for)
+  jobs_to_create = upload_ids_to_adjust_acl_for.count / 100
```
Comment (nbianca) [inline lib/tasks/uploads.rake:605]: I believe this should be rounded up.

File: lib/tasks/uploads.rake [hunk 3]
```diff
@@ -666,45 +659,81 @@ def log_rebake_errors(rebake_errors)
...
-      secure: false,
-      security_last_changed_at: Time.zone.now,
-      security_last_changed_reason: "upload security rake task mark as not secure"
+    UPDATE uploads
+    SET secure = true,
+        security_last_changed_reason = 'upload security rake task mark as secure',
+        security_last_changed_at = NOW()
+    FROM upl
+    WHERE uploads.id = upl.upload_id AND NOT uploads.secure
+    RETURNING uploads.id;
```
Comment (nbianca) [inline lib/tasks/uploads.rake:681]: Nitpick: I noticed you are ending all queries with `;`. I do not think that is necessary and usually we do not do it.

**Review by nbianca — APPROVED** [review]

File: lib/tasks/uploads.rake (hunk 3 above)
Comment (martin-brennan) [reply lib/tasks/uploads.rake:681]: Fair enough, will remove :)

File: lib/tasks/uploads.rake (hunk 2 above)
Comment (martin-brennan) [reply lib/tasks/uploads.rake:605]: Ah yes thank you nice catch!

File: app/jobs/regular/sync_acls_for_uploads.rb (hunk 1 above)
Comment (martin-brennan) [reply app/jobs/regular/sync_acls_for_uploads.rb:22]: Not easily, but it does look like we can use the S3 batch operations / jobs https://docs.amazonaws.cn/en_us/AmazonS3/latest/userguide/batch-ops.html

For example there is one to replace every ACL in the provided manifest (which would be a list of keys) https://docs.amazonaws.cn/en_us/AmazonS3/latest/userguide/batch-ops-put-object-acl.html

This warrants further exploration I think, I will open a topic internally, but it's out of scope for this PR

**Pull request merged**